src/query/util/promqlengine/cache.go

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package promqlengine
+
+import (
+	"time"
+
+	"github.com/prometheus/prometheus/promql"
+)
+
+// NewEngineFn is a function to create a new promql.Engine with given lookback duration.
+type NewEngineFn func(lookbackDelta time.Duration) (*promql.Engine, error)
+
+// Cache caches PromQL engines by lookback duration.
+type Cache struct {
+	cachedEngines map[time.Duration]*promql.Engine
+	newEngineFn   NewEngineFn
+}
+
+// NewCache creates a new Cache.
+func NewCache(newEngineFn NewEngineFn) *Cache {
+	return &Cache{
+		cachedEngines: make(map[time.Duration]*promql.Engine),
+		newEngineFn:   newEngineFn,
+	}
+}
+
+// Get returns a new or cached promql.Engine.
+func (c *Cache) Get(lookbackDelta time.Duration) (*promql.Engine, error) {
+	if engine, ok := c.cachedEngines[lookbackDelta]; ok {
+		return engine, nil
+	}
+
+	return c.newEngineFn(lookbackDelta)
```

Comment:
hmm, so we will be returning new promQL engines on each cache miss? Shouldn't we update the cache in this case?

Reply:
We've discussed this and agreed on this as a trade-off: cache the `promql.Engine`s that have lookbacks based on namespace resolutions, and create the others on the fly. Since setting `&lookback` in request is quite obscure feature, there shouldn't be that many engines created on the fly.

We could try caching all the engines, but then we will have to handle eviction. This would add extra code and there are some inconveniences related to `promql.Engine` metrics. Currently only a single `promql.Engine` can publish metrics, because if you try enabling metrics for subsequent engines, you get errors about duplicate metric registrations. So if you're not careful about what engines are evicted, you may lose metrics or risk duplicate-metric-registration errors.

I guess we could discuss with others if current implementation is sufficient, or if we should make it more sophisticated.

Reply:
Agree that we probably don't want to over-engineer this for now, but I was thinking maybe we could just return some `default` engine in case we don't find cached engine?

Reply:
There doesn't seem anything too heavy to instantiate in `promql.Engine`, it should be ok to create engines on the fly for the rare requests with custom lookback values